Best practices/React/Organization.md

```diff
@@ -4,11 +4,109 @@
 
 This section describes a sensible starting point for organizing a React application that supports server-side rendering.
 
-## `/app`
+## Application
 
-### `/app/components` and `<Component>/components`
+### React Components and Sub-Components
+
+React Components should always exist in a `components/` folder, but the component's internal structure may change depending on their complexity.
+
+
+#### 1. Start with a file and a test
+
+At its simplest a component will exist as a component and a test at the `components/` level.
+
+If you foresee this component growing quickly in complexity or needing a subcomponent, feel free to jump straight to #2.
+```
+.
+└── components
+    ├── ...
+    ├── MyComponent.js
+    └── tests
+        ├── ...
+        └── MyComponent.test.js
+```
+
+
+#### 2. Isolate your component to a folder (of the same name) as the component grows
+
+```
+├── component.js
+└── components
+    └── MyComponent
+        ├── MyComponent.js
+        ├── MyComponent.scss
+        └── MyComponent.test.js
```

Comment:
This should be tucked in a `tests` folder 100% of the time.

Reply:
I would challenge this. Most components I see with a `tests/` folder (in partners at least) have only one test file in them.

I understand that when you have lots of different tests there is a benefit in grouping them in a `tests/` folder (I deliberately used that as my first example 😄). What I would like to see less of is single-file-containing folders, I don't see a benefit of isolating a single test in a folder. Which is why I prefer it as optional, and mention it when grouping the files.

I'm totally open to hearing the reasoning though if you've got the time 🤗

Reply:
Basically should a single test file go inside a folder. I think it's fair to argue against that 👍 I am likely coming at this with a `web`-bias [where](https://github.com/Shopify/web/tree/master/app/foundation/App/tests) [we](https://github.com/Shopify/web/tree/master/app/foundation/App/components/AppDetailsProvider/tests) [always](https://github.com/Shopify/web/tree/master/app/foundation/App/components/Monorail/tests)  [have](https://github.com/Shopify/web/tree/master/app/foundation/App/components/ContentSecurityPolicy/tests)  a `tests` folder, so seeing them run loose with the other files looks strange.

Reply:
Yeah, those are good examples of why I'm pushing back on that convention 😄 

e.g. 
https://github.com/Shopify/web/tree/master/app/foundation/App/components/ContentSecurityPolicy

```
└── components
    └── ContentSecurityPolicy
        ├── ContentSecurityPolicy.tsx
        └── tests/
```
reads just as clean as
```
└── components
    └── ContentSecurityPolicy
        ├── ContentSecurityPolicy.tsx
        └── ContentSecurityPolicy.test.tsx
```
with the exception that in the 2nd example I can see and open the test, and as a bonus I can see there's only one 😉 

But this is good to know 👌